### PR TITLE
[FLINK-27468] Recover missing deployments and other cancel/upgrade improvements for 1.15

### PR DESCRIPTION
**Pull request by gyfora**
This PR contains the following features improvements:

- Add general ability to recover MISSING JobManager deployments (that were deleted by the user or otherwise disappeared )
- Improve Job and JobManager deployment state tracking by moving status handling into the methods that actually interact with the clusters/deployments
- Disable application cluster shutdown for 1.15 in case of failures and savepoint suspend operations in order to be able to track job status for terminal states and avoid savepoint upgrade corner cases
- Improve test coverage by adding extra guards in TestingFlinkService and using the FlinkService implementation as much as possible
- Introduce paremerized tests that cover 1.15 cluster shutdown behaviour
- Introduce new config option to enable deployment recovery (disabled by default in 1.14 enabled 1.15 and onward)
- Replace usage of special UNKNOWN string for jobstate with RECONCILING which allows us to map all states to flink jobstatus easily

Follow up work not covered here:
 - Recover last checkpoint/savepoint information from clustery (https://issues.apache.org/jira/browse/FLINK-27495)
 - Add events for terminal job states and record in deployment status for Flink 1.15 (https://issues.apache.org/jira/browse/FLINK-27497)
 - Add E2E tests to cover Flink 1.15 (https://issues.apache.org/jira/browse/FLINK-27498)
 - Bump base Flink dependency to 1.15 and remove some copied code (https://issues.apache.org/jira/browse/FLINK-27499)
 - Consider making JobStatus.state enum (Flink JobStatus) which is now possible

### CODE REVIEW

**Comment by gyfora** [comment]
cc @tweise @Aitozi @wangyang0918 @morhidi

**Comment by tweise** [comment]
@gyfora one question regarding the motivations to delete a deployment externally: This could be used as escape hatch by users that want to term a misbehaving deployment and have problems removing the CR (which we discussed in the past). I wonder if by default we should or should not try to bring back deployments? And if the default is to try that, what do we advise in the case where a CR update isn't possible (for example due to incompatible CRD update or issue with the operator)?

**Comment by gyfora** [comment]
@tweise hm, thats a good question. 

In general my logic for making the default behaviour to recover (at least from Flink 1.15 onward), is this is how operators usually work. If you delete a resource managed by an operator it is usually brought back up for you. I would consider the jobmanager deployment a managed resource so we should try to do everything to recover if possible.

The only reason I made this default off in Flink 1.14 is because we cannot keep the clusters around after the job terminated which produces a bunch of annoying cases for failed/finished jobs.

As for what to do if your FlinkDeployment CR gets stuck and you cannot remove it, I don't really know :/ 
We should make sure that we do not introduce breaking changes 😄

**Comment by gyfora** [comment]
We could add a config option to blacklist FlinkDeployments that the operator should not touch, that would solve this in a more general way, but it still feels a bit strange to do this.

**Comment by morhidi** [comment]
How about adding some retry count as a guard which is cleared when the user changes the CR?

**Comment by gyfora** [comment]
@tweise 
After discussing this in detail with @morhidi we came to the following conclusions:

If there is a CR incompatibility issue, the operator won't be able to reconcile it anyways so the recovery logic (or anything else in the controller) won't take effect so the user can simply delete the jobmanager deployment.

Otherwise if the user wants to stop the job they simply need to suspend the FlinkDeployment. The operator should do anything in it's power to make sure that the Flink/Cluster state reflects what's in the FlinkDeployment, therefore automatic jm deployment recovery should be the default in 1.15 and later.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java [hunk 1]
```diff
@@ -255,4 +259,45 @@ public static boolean shouldRollBack(
...
+                conf.getOptional(
+                                KubernetesOperatorConfigOptions
+                                        .OPERATOR_RECOVER_JM_DEPLOYMENT_ENABLED)
+                        .orElse(
+                                conf.get(FlinkConfigBuilder.FLINK_VERSION)
+                                                .isNewerVersionThan(FlinkVersion.v1_14)
+                                        ? true
+                                        : false);
+
+        return enabled
```
Comment (Aitozi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java:275]: Will this also trigger a redeployment for a batch job finished ?

Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java:275]: Not in 1.15 as the cluster won't shut down (so the jobmanager deployment is not lost). In flink 1.14 if you enable this it would redeploy the batch job, this is exactly the reason why this is disabled in 1.14 by default.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java [hunk 2]
```diff
@@ -211,6 +219,22 @@ protected boolean isSuspendedJob(FlinkDeployment deployment) {
                 && lastReconciledSpec.getJob().getState() == JobState.SUSPENDED;
     }
 
+    private void onMissingDeployment(FlinkDeployment deployment) {
```
Comment (Aitozi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java:222]: Can we distinguish the `onMissingDeployment` cause here (like by judge whether there is an on-going upgrade). If it caused by Operator, we can continue reconcile. Otherwise we can just clean up the `FlinkDeployment` which means an external operation or internal finish/failed happened.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java (hunk 1 above)
Comment (Aitozi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java:275]: Get it

Comment (Aitozi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java:275]: Get it

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java (hunk 2 above)
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java:222]: On-going upgrade would never trigger this , when the job is in suspended state with missing deployment we won't try to observe it. 

Even if we trigger there is a problem we should never clean up the deployment, as the user probably needs to manually recover the job. Cleanup is reserved for deletions.

**Comment by wangyang0918** [comment]
I am getting to this PR and try to give it a pass very soon.

**Comment by gyfora** [comment]
@wangyang0918 let me know if you have any questions, it's not a trivial change but there are a lot of big improvements here I believe 😄

**Comment by morhidi** [comment]
nit: Job status successfully updated from RECONCILING to Optional[RUNNING] can you please fix the logging and get the value from the Optional?

**Comment by morhidi** [comment]
I tested the basic functionality on my Minikube, feels very K8s native :) Great feature!

**Review by wangyang0918 — COMMENTED** [review]
@gyfora You are right. This is not a trivial change and I have gone through this PR carefully.

Except for the existing, I have one more comment about the metrics. If the flink job failed and `SHUTDOWN_ON_APPLICATION_FINISH` is false, then `JobManagerDeploymentStatus` is still `READY`. So we might need some metrics about the job status. cc @morhidi

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java [hunk 3]
```diff
@@ -255,4 +259,40 @@ public static boolean shouldRollBack(
...
+    public static boolean isJobInTerminalState(FlinkDeploymentStatus status) {
+        JobManagerDeploymentStatus deploymentStatus = status.getJobManagerDeploymentStatus();
+        if (deploymentStatus == JobManagerDeploymentStatus.MISSING) {
+            return true;
+        }
+
+        String jobState = status.getJobStatus().getState();
+
+        return deploymentStatus == JobManagerDeploymentStatus.READY
+                && org.apache.flink.api.common.JobStatus.valueOf(jobState).isTerminalState();
```
[the remark this answers is not in the document]
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java:291]: The only LOCALLY terminal state is SUSPENDED. Which we can also use for updates. So I think `isTerminalState` is correct here.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java [hunk 4]
```diff
@@ -211,6 +220,22 @@ protected boolean isSuspendedJob(FlinkDeployment deployment) {
...
+        String err = "Missing JobManager deployment";
+        logger.error(err);
```
[the remark this answers is not in the document]
Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java:225]: MDC adds this I guess, or you meant something else?

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconciler.java [hunk 5]
```diff
@@ -173,6 +174,16 @@ private void rollbackApplication(FlinkDeployment flinkApp) throws Exception {
                 flinkApp.getMetadata(), rollbackSpec, rollbackConfig, kubernetesClient);
     }
 
+    private void recoverJmDeployment(FlinkDeployment deployment) throws Exception {
+        LOG.info("Missing Flink Cluster deployment, trying to recover...");
```
[the remark this answers is not in the document]
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconciler.java:178]: the name is included in the MDC already

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (wangyang0918) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java:291]: IIUC, the `SUSPENDED` is not a persistent state and it will transit to other states(e.g. `RUNNING`, `FAILED) finally. For example, the JobManager lost leadership and gain back.

[the remark this answers is not in the document]
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java:291]: Hm I see, I wasn't sure about that. I can change to globallyTerminal

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java [hunk 6]
```diff
@@ -112,30 +132,31 @@ protected FlinkConfigBuilder applyFlinkConfiguration() {
...
-                effectiveConfig.set(
+            if (spec.getJob().getUpgradeMode() == UpgradeMode.LAST_STATE) {
+                setDefaultConf(
                         ExecutionCheckpointingOptions.CHECKPOINTING_INTERVAL,
                         DEFAULT_CHECKPOINTING_INTERVAL);
             }
+
+            // We need to keep the application clusters around for proper operator behaviour
+            effectiveConfig.set(SHUTDOWN_ON_APPLICATION_FINISH, false);
+            if (HighAvailabilityMode.isHighAvailabilityModeActivated(effectiveConfig)) {
```
[the remark this answers is not in the document]
Comment (wangyang0918) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java:154]: I got it. Multiple jobs could be submitted when HA disabled.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/JobStatusObserver.java [hunk 7]
```diff
@@ -57,7 +55,7 @@ public boolean observe(JobStatus jobStatus, Configuration deployedConfig, CTX ct
             clusterJobStatuses = new ArrayList<>(flinkService.listJobs(deployedConfig));
         } catch (Exception e) {
             LOG.error("Exception while listing jobs", e);
-            jobStatus.setState(JOB_STATE_UNKNOWN);
+            jobStatus.setState(org.apache.flink.api.common.JobStatus.RECONCILING.name());
```
[the remark this answers is not in the document]
Comment (wangyang0918) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/JobStatusObserver.java:58]: We could keep it as now and change later if necessary.

**Comment by wangyang0918** [comment]
If you do not mind, I will do some manual tests tomorrow and share the feedback then.

**Comment by gyfora** [comment]
@wangyang0918 of course, I really appreciate it

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java [hunk 8]
```diff
@@ -211,6 +220,22 @@ protected boolean isSuspendedJob(FlinkDeployment deployment) {
...
+        Event event =
+                DeploymentFailedException.asEvent(
+                        new DeploymentFailedException(
+                                DeploymentFailedException.COMPONENT_JOBMANAGER, "Error", err),
+                        deployment);
+        kubernetesClient
```
[the remark this answers is not in the document]
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java:239]: I am not really sure how to fix this one

[the remark this answers is not in the document]
Comment (wangyang0918) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java:239]: I think it is due to `lastTransitionTime` and `lastUpdateTime` is null.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (wangyang0918) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java:233]: My bad. We might not need to add namespace and name here. MDC add this. We also do not need this for the event. It is duplicated.

```
wangyang-pc:flink-kubernetes-operator danrtsey.wy$ kubectl get events | grep 'flinkdeployment/flink-example-statemachine'
11h         Error     Missing JobManager deployment for default/flink-example-statemachine   flinkdeployment/flink-example-statemachine         Missing JobManager deployment for default/flink-example-statemachine
```

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java [hunk 9]
```diff
@@ -255,4 +259,50 @@ public static boolean shouldRollBack(
                 .minus(readinessTimeout)
                 .isAfter(Instant.ofEpochMilli(reconciliationStatus.getReconciliationTimestamp()));
     }
+
+    public static boolean deploymentRecoveryEnabled(Configuration conf) {
+        return conf.getOptional(
+                        KubernetesOperatorConfigOptions.OPERATOR_RECOVER_JM_DEPLOYMENT_ENABLED)
+                .orElse(
+                        conf.get(FlinkConfigBuilder.FLINK_VERSION)
```
Comment (wangyang0918) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java:267]: nit: `? true : false` is unnecessary.

**Comment by wangyang0918** [comment]
I have done some manual tests. And it works well for recover the missing deployment. I have last one question which need to be discussed more.

When `SHUTDOWN_ON_APPLICATION_FINISH` is configured to false, then it is the operator's responsibility to clean up the K8s resources. However, I do not find such logics. It means that if the job reached a globally terminal state(e.g. failed, finished) or suspended manually(changing the CR `.spec.job.state`), we will have the residual K8s resources. In my opinion, the operator needs to do the clean-up after fetching the job status and store it into the CR status successfully.

**Comment by gyfora** [comment]
@wangyang0918 If I understand correctly, in these cases only the jobmanager deployment remains but the taskmanagers are released.
The jobmanager deployment will be deleted either on next upgrade or when the user deletes the FlinkDeployment itself.

I think this works quite well, it allows users to go to the Flink cluster, inspect logs, configs, metrics etc. And since the heavy resources are already gone it won't put a strain on the cluster.

**Comment by wangyang0918** [comment]
Your explanation makes sense to me. Given that it is the designed behavior, I have no more question now.

I just feels like if the job finished(e.g. batch jobs) successfully or suspended manually, we do not need to have the retained jobmanager deployment. Anyway, the users could delete the CR to trigger the deletion.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java:238]: Pushed a fix for this:
image

**Review by wangyang0918 — APPROVED** [review]
+1 for merging.
